client/sniffer/src/main/java/org/elasticsearch/client/sniff/Sniffer.java

```diff
@@ -106,35 +119,63 @@ synchronized void scheduleNextRun(long delayMillis) {
 
         @Override
         public void run() {
-            sniff(null, sniffIntervalMillis);
-        }
-
-        void sniffOnFailure(HttpHost failedHost) {
-            sniff(failedHost, sniffAfterFailureDelayMillis);
+            sniff(sniffIntervalMillis);
         }
 
-        void sniff(HttpHost excludeHost, long nextSniffDelayMillis) {
+        void sniff(long nextSniffDelayMillis) {
             if (running.compareAndSet(false, true)) {
+                long nextSniffDelay = nextSniffDelayMillis;
                 try {
                     List<HttpHost> sniffedHosts = hostsSniffer.sniffHosts();
                     logger.debug("sniffed hosts: " + sniffedHosts);
-                    if (excludeHost != null) {
-                        sniffedHosts.remove(excludeHost);
-                    }
-                    if (sniffedHosts.isEmpty()) {
+
+                    List<HttpHost> hostsFiltered = removeExcludedAndCycle(sniffedHosts);
+                    logger.debug("sniffed hosts after filtering: " + sniffedHosts);
+
+                    if (hostsFiltered.isEmpty()) {
                         logger.warn("no hosts to set, hosts will be updated at the next sniffing round");
                     } else {
-                        this.restClient.setHosts(sniffedHosts.toArray(new HttpHost[sniffedHosts.size()]));
+                        this.restClient.setHosts(hostsFiltered.toArray(new HttpHost[hostsFiltered.size()]));
                     }
                 } catch (Exception e) {
                     logger.error("error while sniffing nodes", e);
+                    nextSniffDelay = sniffAfterFailureDelayMillis;
```

Comment:
The idea behind `sniffAfterFailureDelay` is that once a node fails, you want to re-sniff, but you may also want to check if the node has come back through a new sniffing round earlier than you'd usually do. So in general `sniffAfterFailureDelay` would be lower than `sniffInterval`. I see that here you are using now a different interval for cases where sniffing itself has failed, but that was not the purpose of the `sniffAfterFailureDelay`. It is not a bad idea to actually re-sniff earlier if something goes wrong, but it should be a different concept.